%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                                                 %%
%% Please do not use \input{...} to include other tex files.       %%
%% Submit your LaTeX manuscript as one .tex document.              %%
%%                                                                 %%
%% All additional figures and files should be attached             %%
%% separately and not embedded in the \TeX\ document itself.       %%
%%                                                                 %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\documentclass[sn-mathphys,Numbered, lineno]{sn-jnl}  % I added lineo to have linenumbers

%%%% Standard Packages
%%<additional latex packages if required can be included here>
\usepackage{graphicx}%
\usepackage{multirow}%
\usepackage{amsmath,amssymb,amsfonts}%
\usepackage{amsthm}%
\usepackage{mathrsfs}%
\usepackage[title]{appendix}%
\usepackage{xcolor}%
\usepackage{textcomp}%
\usepackage{manyfoot}%
\usepackage{booktabs}%
\usepackage{algorithm}%
\usepackage{algorithmicx}%
\usepackage{algpseudocode}%
\usepackage{listings}%

%%%%
% REMOVE THOSE IN THE END
\usepackage{todonotes}
\usepackage{caption}
\usepackage{tabularx}
\usepackage{adjustbox}  % for the adjustwidth



\usepackage{threeparttable}  % to support notes within a table



%\jyear{2021}%

\theoremstyle{thmstyleone}%
\newtheorem{theorem}{Theorem}%  meant for continuous numbers
\newtheorem{proposition}[theorem]{Proposition}% 
\theoremstyle{thmstyletwo}%
\newtheorem{example}{Example}%
\newtheorem{remark}{Remark}%
\theoremstyle{thmstylethree}%
\newtheorem{definition}{Definition}%
\raggedbottom


\newcommand{\microbetag}{\texttt{microbetag }}


\begin{document}

\title[Article Title]{
    % Genomic, metabolic and literature oriented annotation of microbial co-occurrence networks enhances associations confidence level and hypothesis generation
    microbetag: simplifying microbial network interpretation through annotation, enrichment and metabolic complementarity analysis
}


%%==================================%%
%% Authors - affiliations           %%
%%==================================%%


\author[1]{\fnm{Haris} \sur{Zafeiropoulos}}\email{haris.zafeiropoulos@kuleuven.be}
\author[1]{\fnm{Ermis Ioannis} \sur{Michail Delopoulos}}\email{ermisioannis.michaildelopoulos@student.kuleuven.be}
\author[2]{\fnm{Andi} \sur{Erega}}\email{andi.erega@hest.ethz.ch}
\author[2]{\fnm{Annelies} \sur{Geirnaert}}\email{annelies.geirnaert@hest.ethz.ch}
\author[3]{\fnm{John} \sur{Morris}}\email{scooter@cgl.ucsf.edu}
\author*[1]{\fnm{Karoline} \sur{Faust}}\email{karoline.faust@kuleuven.be}


\affil*[1]{
    \orgdiv{
        Department of Microbiology, Immunology and Transplantation, Rega Institute for Medical Research
    }, 
    \orgname{KU Leuven}, 
    \orgaddress{
        \street{Herestraat}, \city{Leuven}, \postcode{3000}, \state{}, \country{Belgium}
    }
}
\affil[2]{
    \orgdiv{Institute of Food, Nutrition and Health}, 
    \orgname{ETH Zurich}, 
    \orgaddress{
        \street{Street}, \city{Zurich}, \postcode{8092}, \state{}, \country{Switzerland}
    }
}
\affil[3]{
    \orgdiv{Department of Pharmaceutical Chemistry}, 
    \orgname{University of California San Francisco}, 
    \orgaddress{
        \street{Street}, \city{San Francisco}, \postcode{94143}, \state{California}, \country{USA}
    }
}


%%==================================%%
%% unstructured abstract %%
%%==================================%%
\abstract{

    \footnote{
        Looks liks Chris Quince is~\href{https://microbiomejournal.biomedcentral.com/articles/sections/bioinformatics--algorithms-and-software}{our editor}.
    }

    Up to 350 words.

    The abstract must include the following separate sections:
    
    \textbf{Background:} the context and purpose of the study

    \textbf{Results:} the main findings

    \textbf{Conclusions:} a brief summary and potential implications


    \begin{figure}[H]
        \label{fig:abstract}
        \includegraphics[width=\columnwidth]{figs/figure_abstract.png}
        \caption*{
            Figure abstract. 
            
            % From a count table (bins or OTUs/ASVs) one can come up with a co-occurrence network. 
            % To better understand and assess the confidence level of these associations, microbetag annotates both taxa (nodes) 
            % and 
            % integrating 
            % genomic data 
            
        }
    \end{figure}
    

}
%%==================================%%
%% keywords                         %%
%%==================================%%
\keywords{
    microbial associations, 
    enrichemnt analysis, 
    data integration,
    pathway complementarity,
    seed set
}



\maketitle


%%==================================%%
%% MAIN BODY                        %%
%%==================================%%


% ====================================================
% Background
% ====================================================
% REMEMBER: To have a footnote in the \section{}, you need to have NO empty lines; otherwise latex will not compile
\section*{Background
    \footnote{
        We are to submit in the Microbiome journal as a "Software" manuscript, thus we follow
        ~\href{https://microbiomejournal.biomedcentral.com/submission-guidelines/preparing-your-manuscript/software-article}{these rules}
    }
    \footnote{
        The introduction should not include subheadings.
        The Background section should explain the relevant context and the specific issue that the software described is intended to address.
        }
}
\label{sec:background}

    Microbial ecology plays a fundamental role in the stability and resilience of ecosystems and their processes; from soils, aquatic environments and biogeochemical cycles~\cite{yuan2021climate} to host-associated environments and the human health~\cite{raes2008molecular, faust2012microbialReviewInteractions}.
    Most microbial species live only in communities~\cite{rottjers2018hairballs} and most natural microbial communities consist of hundreds or even thousands of species~\cite{balint2016millions}.
    Each species exhibits a unique repertoire of reactions and adapts to various niches, each with specific nutrient and environmental requirements.
    % KF: "To unravel microbial ecology involves disentangling the principles that dictate the organization of a community"... sounds rather pompous, can you tone it down a bit? we don't really look at the principles of community organisation in this work.
    % HZ: What about now ?
    Understanding the dynamics governing interactions among microbial species and their relationships with the surrounding environment would sheld light in several aspects microbial ecology~\cite{robinson2010structure}.
    % references if we'd like to talk about microb com as ~\textit{complex systems}: ~\cite{godfray1990complex, amor2017spatial},

    % KF: Sentence below is quite complex; is there an easier way to say this? Also, what are metabolic niche effects? Be careful with big words like "paradigm" - is this word appropriate here?
    % HZ: What about now ? 
    % In recent years, there has been a compelling suggestion proposing an elegant paradigm for microbial ecology: 
    Metabolic conditions within an environment influence the composition and distribution of its functions (metabolic niche effect).
    At the same time, processes like genetic diversity and stochastic events, induce variability within different functional groups of the community.
    Recent studies suggest that the relative importance of those two processes drive the extent to which functional and taxonomic composition at the community level are correlated~\cite{louca2016high}.
    %the correlation between community functional and taxonomic composition depends on the relative importance of metabolic niche effects 
    %% (influence of specific metabolic conditions and requirements within an environment on the composition and distribution of microbial functions)
    %compared to processes causing variation within functional groups 
    %% (genetic diversity, regulatory mechanisms, environmental influences, and stochastic events affecting individual members of the functional group)
    % KF: what is the link of all this to microbetag? I suggest to strongly reduce this part and write instead more about interaction mechanisms and why it is so difficult to infer interactions from abundance data
    +++++++++
    The notion of an interaction varies including cooperation, competition, parasitism, commensalism and ammensalism~\cite{faust2012microbialReviewInteractions}.

    % % ----------   Taxonomy - functional composiiton relationship ---------------
    % In this case, comparable environments should foster similar microbial community functions, even though there may be taxonomic variations within individual functional groups, 
    % while in more heterogeneous environments functional $\beta$ diversity would be strongly correlated with taxonomic $\beta$ diversity.
    % In the latter, the decoupling between community composition and metabolic functioning is concealed by robust metabolic niche effect~\cite{louca2016high}.
    % Thus, an interaction between two taxa may vary in different environments.
    % Microbial interactions can be the result of multiple phenomena, such as exchange of metabolic products~\cite{kost2023metabolic}, biofilm formation~\cite{arnaouteli2021bacillus}, gene transfering~\cite{de2023horizontal} and signaling~\cite{keller2006communication}.
    % % The metabolic function of a community undergoes substantial influence from energetic and stoichiometric constraints~\cite{george2023functional}.

    High-throughput sequencing (HTS) has provided great insight into the diversity and composition of microbial communities~\cite{elixir_microbiome}. % get more refs
    Uncultivated species can now be detected and their featues can be inferred through their genomic information~\cite{hug2016new}.
    Moreover, the composition of thousands of microbiome samples is now accessible allowing for the inference of patterns among sets of samples.
    A widely used approach to extraxt such patterns, is the creation of co-occurrence networks based on metagenomic read data (amplicon and/or shotgun)~\cite{matchado2021network}. 
    A great number of approaches is available for co-occurrence network inference based on a range of statistical concepts such as: correlation (e.g., CoNet~\cite{faust2012microbial}, SparCC~\cite{friedman2012inferring}), linear regression (e.g., SpiecEasi~\cite{kurtz2015sparse}) and causal inference (FlashWeave~\cite{flashweave_cite}).
    Nevertheless, microbial co-occurrence networks continue to encounter various challenges~\cite{faust2021open}, encompassing issues associated with data analysis and network construction, leading to tool-dependent analysis~\cite{kishore2023inferring, weiss2016correlation, rottjers2018hairballs}.
    % KF: verb missing in the sentence below
    But also, challenges regarding the interpretation of the networks.
    Addressing the question of~\textit{What can we learn from the hairballs} posed by R{\"o}ttjers et \textit{al.}~\cite{rottjers2018hairballs} could provide essential insight on the mechanisms of the interactions.

    % KF: why has the use of network inference for interaction prediction underscored its limited accuracy? Imprecise expression: the use itself did not underscore the low accuracy - evaluations did. This needs to be rephrased.
    The use of microbial network inference as a means for predicting interactions has underscored its limited accuracy, and the fact that the biological implications of network properties remain unclear~\cite{berry2014deciphering}.
    Theoretical principles derived from network studies might provide indications of emergent biological characteristics~\cite{rottjers2018hairballs, guo2022microbial}. 
    For example, modules (highly interconnected nodes) within microbial co-occurrence networks could serve as indicators of ecological processes that govern community structure, including niche filtering and habitat preference\cite{ma2020earth}.
    Data integration and clustering have been suggested to address this challenge~\cite{faust2021open}.
    Clusters identified in microbial association networks have demonstrated their ability to mirror key drivers of community composition~\cite{guidi2016plankton} and sevaral algorithms and implementations are available~\cite{rottjers2020manta}.
    However, data integration approaches in microbial co-occurrence networks are so-far limited.
    Here, we present \microbetag, a microbial co-occurrence network annotator that exploits several channels of information to enhance/diminish the confidence of the associations suggested by the network and generate hypotheseses for further investigation both at the taxon pair and the community level.

    \microbetag serves as a comprehensive platform that provides information on taxa along with their potential metabolic interactions from multiple channels 
    (see Implementation~\ref{sec:implementation}).
    % literature-based classification of OTUs and their corresponding taxonomies to functional groups, 
    % prediction of microbial phenotypes via comparative genomics
    The key concept here is the reverse ecology approach~\textit{reverse ecology}~\cite{levy2012reverse}.
    Reverse ecology leverages genomics to explore community ecology with no \textit{a priori} assumptions about the taxa involved.
    Making the most of advancements in systems biology and genomic metabolic modeling, as well as system-level analysis of intricate biological networks, the reverse ecology framework enables the prediction of ecological traits for less-understood microorganisms, their interactions with others, and the overall ecology of microbial communities~\cite{levy2014metagenomic}.
    % KF: what was major about the seed set analysis? Did we learn anything useful from it so far? I don't think so, so please tone it down.
    % HZ: I suggest we just remove this. 
    % In this context, seed set analysis has been a major contribution in the study of both the species and the community ecology based on their genetic information.

    A metabolic network's "seed set" is the set of compounds that, based on the network topology, need to be acquired exogenously~\cite{borenstein2008large}~(see Figure~\ref{fig:pathcompl}).
    Such nodes might be independent, i.e. they cannot be activated by any other node in the network, or they can be interdependent forming groups of seed nodes.
    Seeds are a useful proxy for the habitat of the organism and an essential tool in the frameowrk of reverse ecology~\cite{parter2007environmental,borenstein2008large}.
    Based on the seed concept, several graph theory-based metrics (indices) have been described to predict species interactions directly from their networks' topologies~\cite{levy2015netcooperate, kreimer2012netcmpt, zelezniak2015metabolic, belcour2020metage2metabo}.
    Over the last years, the seed apprach has been implemented at the Genome-scale metabolic network reconstructions (GENREs) level.
    GENREs encapsulate mathematical representations capturing the biochemical reactions that could take place within an organism~\cite{thiele2010protocol, durot2008genome, communityGemMicrobiomeModel2023}.

    Metabolic complementarity among species, serving as a reflection of potential cooperation within communities, assesses the capacity for collaboration; cross-feeding or syntrophy interactions are typical examples of such a collaboration. 
    In contrast, metabolic competition refers to the metabolic overlap between two species leading to exploitative competition, e.g. for nutrient resources.
    Seed and non-seed sets can be used to compute such indices. 
    Thorough examination of such complements can reveal metabolic interactions leading to patterns observed on the co-occurrence network.
    
    % KF: In addition, I don't get the meaning of the first part of this sentence
    % HZ: 
    Considering complementarity as a range of alternatives and pairwise microbial interactions in the context of the community as a whole, microbial species may also exchange metabolic compounds that may be not seeds at a certain time but may allow them to perform functions that they are currently not capable of~\cite{mori2016metabolic,zientz2004metabolic}.
    Such by-products may be even metabolites not even necessary for themeselves but for the community as a whole~\cite{kallus2017paradoxes}.
    To explore the potenial of a species metabolism given they benefit from a partner of theirs, genome annotations combined with collections of functional units to highlight can provide a valid proxy.
    We present here a naive approach exporting all possible complements between a pair of species based on their KEGG ORTHOLOGY (KOs) annotations and the KEGG MODULES database~\cite{keggmodulesdb}.

    \microbetag annotates a user's co-occurrence network by integrating phenotypic traits on the taxa present on the network (nodes) and potential metabolic interactions to their suggested associations (edges).
    A Graphical User Interface (GUI) is supported as a CytoscapeApp providing a user-friendly environment to investigate annotations in a straightforward way.
    All annotations present in microbetagDB are also available through an Application Programming Interface (API).
    \microbetag's source code is distributed under a GNU GPL v3 license and available on GitHub. 
    Documentation and further support on how to use \microbetag is available at \href{https://hariszaf.github.io/microbetag/}{documentation web-site}.
    To the best of our knowledge there is not a software with which \microbetag could be compared with directly. 
    To validate our annotations we used a recently published network with partially known interactions between some pairs of species found associated~\cite{hessler2023vitamin} (see Results section, paragraph~\ref{subsec:validation}).    
    To demonstrate \microbetag's potential, we present the main features of its interface and we discuss a real-world use-case (see Discussion section, paragraph~\ref{subsec:usecase}).


    % ===================           consider to be use if necessary         =====================
    % From~\cite{louca2016high}
    % Ecological communities are governed by an interplay of deterministic processes associated with species interactions with their environment (environmental filtering) and each other (limiting similarity)1,2.

    % the metabolic functional potential of microbial communities in the global ocean or in soil is closely related to environmental conditions, while the taxonomic variation within individual functional groups is only poorly explained by environmental condition

    % community metabolic function is strongly shaped by energetic and stoichiometric constraints such as the availability of electron acceptors for respiration10, while the composition within functional groups is modulated by additional mechanisms.

    % According to this paradigm, \textbf{similar environments should promote similar microbial community function, while allowing for taxonomic variation within individual functional groups}.

    % These interactions contribute to the overall stability and functioning of ecosystems by influencing nutrient cycling, disease dynamics, and the overall diversity and composition of microbial communities. 
    % For example, some microorganisms can produce and release certain compounds that benefit neighboring microbes or inhibit the growth of competing species. 
    % Other interactions involve symbiotic relationships, where different microbes rely on each other for survival and perform complementary functions. 
    % Understanding and studying microbial interactions is vital for unraveling the complexity of microbial communities and their impact on human health, agriculture, and environmental processes.

    % metabolic interactions specifically, microorganisms often engage in metabolic exchanges during their interactions. 
    % For instance, one microbe may produce metabolites that serve as nutrients or signaling molecules for another microbe. These metabolic interactions can involve the transfer of essential nutrients, the breakdown of complex compounds, or the production of secondary metabolites with antimicrobial properties.
    % Overall, understanding the different types of microbial interactions, including metabolic interactions, provides insights into the complexity and dynamics of microbial communities and their impact on various processes in nature.


 
% ====================================================
% Implementation
% ====================================================
\section*{Implementation
    \footnote{
        This should include a description of the overall architecture of the software implementation, along with details of any critical issues and how they were addressed.
    }
}
\label{sec:implementation}

    \subsection*{Genomes included}
    \label{subsec:genomes}

        Using the Genome Taxonomy Database (GTDB) v207 \href{https://data.gtdb.ecogenomic.org/releases/release207/207.0/}{metadata files}, we retrieved the NCBI genome accessions of the high quality representative genomes, i.e. completeness $\geq 95\%$  and contamination $\leq 5\%$.
        % That resulted a set of $26,778$ covering $22,009$ unique NCBI Taxonomy Ids.
        A set of $26,778$ genomes was obtained, representing $22,009$ unique NCBI Taxonomy Ids.
        Using these accession numbers, we were able to download their corresponding \texttt{.faa} files when available 
        % (\href{https://github.com/hariszaf/microbetag/blob/develop/microbetagDB/mappings/gtdb_ncbi/get_gtdb_faa.py}{\texttt{get\_gtdb\_faa.py}}) 
        leading to a set of $16,900$ amino acid sequence files.
        The latter were annotated and used to obtain potential pathway complementarities beween pairs of genomes (see paragraph~\ref{subsec:path-compl}).
        Last, when available, their corresponding annotations on PATRIC database~\cite{wattam2017improvements} were retrieved to reconstruct GENREs~(see paragraph~\ref{subsec:seeds}).


    \subsection*{Taxonomy schemes}
    \label{subsec:taxonomies}

        \microbetag maps the taxonomy of each entry in the abundance table to their corresponding NCBI Taxonomy Id and, if available, their closest GTDB representative genome(s), since several GTDB representative genomes may map to the same NCBI Taxonomy Id.
        Two well established taxonomy schemes are supported: the GTDB~\cite{parks2022gtdb} that is being broadly used for bins and/or MAGs taxonomical classification and the Silva database~\cite{quast2012silva} that is widely used in amplicon studies. Both taxonomy schemes link their taxonomies to NCBI Taxonomy Ids~\cite{schoch2020ncbi}.
        In case none of those two taxonomies was used and the abundance table contains less than 1,000 taxa, \microbetag maps the user provided taxonomies to NCBI Taxonomy. %However, this is done only for abundance tables containing less the 1,000 taxa. 
        To this end, \microbetag makes use of the~\href{https://github.com/seatgeek/thefuzz}{\texttt{fuzzywuzzy}} library that implements the Levenshtein Distance Metric to get the closest NCBI taxon name and thus its corresponding NCBI Taxonomy Id; a relatively high similarity score is used (90) to avoid false positives. 
        Also, using the nodes dump file of NCBI Taxonomy, \microbetag may retrieve the child taxa of a taxon in user's data, along with their corresponding NCBI Taxonomy Ids, if requested by the user.
        % Thus, \microbetag maps a network taxon to their corresponding NCBI Taxonomy Id and if available their corresponding GTDB genomes.
        If the user provides their abundance table with taxonomies already mapped to the GTDB taxonomy, \microbetag will report the best possible annotations in a time efficient manner. 
        % as it is both time efficient and, mainly, it allows for the best annotations \microbetag could report, it is strongly suggested that both amplicon and shotgun-oriented data to be mapped to GTDB taxonomies.
        
        % Having a taxonomy scheme of GTDB or Silva is mandatory for networks with more than $1,000$ taxa.
        
        % The corresponding taxonomy will be mentioned as "microbetag\_prep".
        
        % There is a great number of taxonomies that are being used in such studies, e.g. Silva~\cite{quast2012silva}, Ribosomal Database Project (RDP)~\cite{cole2014ribosomal}, manually curated ones and more, 
        % As a consequence, there is not a standardised format of the taxonomies assigned, from bioinformatics pipelines used for the analysis of such data.


    \subsection*{Network inference }
    \label{subsec:net-infer}

        When a co-occurrence network is not provided by the user, \microbetag exploits FlashWeave~\cite{flashweave_cite} to build one on the fly.
        Yet, \microbetag supports the annotation of networks built from any algorithm/software, in any format Cytoscape can load.

        % a computational approach based on a flexible Probabilistic Graphical Model framework that integrates metadata and predicts direct microbial interactions from heterogeneous microbial abundance data sets with hundreds of thousands of samples. 
        % A flexible Probabilistic Graphical Model framework is used in a computational approach that 
        % incorporates metadata and predicts direct microbial interactions. 
        % This is done using heterogeneous microbial abundance datasets consisting of hundreds of thousands of samples.

        %     •    heterogeneous - enable heterogeneous mode for multi-habitat or -protocol data with at least thousands of samples (FlashWeaveHE)
        %     •    sensitive - enable fine-grained associations (FlashWeave-S, FlashWeaveHE-S), sensitive=false results in the fast modes FlashWeave-F or FlashWeaveHE-F
        %     •    FDR - perform False Discovery Rate correction (Benjamini-Hochberg method) on pairwise associations
        %     •    n_obs_min - don't compute associations between variables having less reliable samples (i.e. non-zero if heterogeneous=true) than this number. -1: automatically choose a threshold.



    \subsection*{\microbetag pre-processing }
    \label{subsec:preprocess}

        In order to aid the user to map their sequences to the GTDB taxonomy, DADA2-formatted 16S rRNA gene sequences for both bacteria and archaea~\cite{ali_alishum_2022_6655692} were used to trained the TAXID classifier of the DECIPHER package~\cite{{murali2018idtaxa}} and are available through the~\href{hub.docker.com/repository/docker/hariszaf/microbetag_prep/}{microbetag preprocess Docker image}.
        Likewise, when the abundance table consists of more than $1,000$ taxa, providing a network as an input is mandatory. Again, to help the user, \microbetag preprocess Docker image supports the inference of a network using FlashWeave. 


    \subsection*{Literature based nodes annotation }
    \label{subsec:fapro}

        Using a set of Tara Ocean samples~\cite{sunagawa2015structure} FAPROTAX~\cite{louca2016decoupling} estimates the functional potential of the bacterial and archaeal communities, by classifying each taxonomic unit into functional group(s) based on current literature, descriptions of cultured representatives and/or manuals of systematic microbiology. 
        In this manually curated approach, a taxon is associated with a function if and only if all the cultured species within the taxon have been shown to exhibit that function. 
        % Apparently, FAPROTAX does not support the annotation of clades with no cultured representatives that have been rising thanks to high throughput sequencing technologies.
        % Further, a taxon could be annotated with more than one functions while functional groups could be nested.
        In its current version, FAPROTAX includes more than 80 functions based on ~7600 functional annotations and covering more than 4600 taxa.
        Contrary to gene content based approaches, e.g. PICRUSt2~\cite{douglas2020picrust2}, FAPROTAX  estimates metabolic phenotypes based on experimental evidence. 

        \microbetag invokes the accompanying script of FAPROTAX and converts the taxonomic microbial community profile of the samples included in the user's abundance table or of the taxa present in the provided network, into putative functional profiles.
        Then, it parses FAPROTAX's subtables to annotate each taxonomic unit present in the user's data with all the functions for which they had a hit. 
        FAPROTAX annotations are not part of the microbetagDB but are computed on the fly.


    \subsection*{Genomic based nodes annotation }
    \label{subsec:phen}

        phenDB~\cite{feldbauer2015prediction} is a publicly available resource that supports the analysis of bacterial (meta)genomes to identify 47 distinct functional traits, e.g. whether a species is producing butanol or has an halophilic lifestyle. 
        It relies on support vector machines (SVM) trained with manually curated datasets based on gene presence/absence patterns for trait prediction.
        More specifically, the model for a particular trait is trained using a collection of EggNOG annotated genomes where the knowledge of whether that trait is present or absent among its members is available.
        These models (classifiers) are used to predict presence/absence of their corresponding traits in non-studied species. 
        % Based on the completeness/contamination of the genomes, the accuracy varies. 
        % The~\texttt{compute-genotype} program of phenotrex supports the creation of such tabular~\textit{genotype} files.
        % A~\textit{genotype} file can be used along with a~\textit{phenotype} one, i.e., a file containing true phenotypic trait values for each input genome on which to train the model, and the~\texttt{train} program of phenotrex can then be performed. 

        % one can implement tests to evaluate the model based on the completeness of a genomes and its contamination (Performance Estimation) using the cccv program. e.g
        % {
        % 	"mean_balanced_accuracy": 0.5499999999999999,
        % 	"stddev_balanced_accuracy": 0.10482201257840669,
        % 	"contamination": 0.7,
        % 	"completeness": 1.0
        % },
        In the framework of microbetagDB, %phenotrex 
        classifiers were re-trained using the genomes provided by phenDB for each trait to sync with the latest version of eggNOG~\cite{huerta2019eggnog} and the~\texttt{phenotrex}~\cite{feldbauer2015prediction} software tool.
        Genomes were downloaded from NCBI using the \href{https://www.ncbi.nlm.nih.gov/sites/batchentrez}{Batch Entrez} program.
        % as there was a conflict between the eggNOG version phenotrex uses at the prediction step and the one during the training of the classes as provided in the PhenDB site.
        % For example, for the acetic acid production case, the corresponding webpage of phenDB pointed to the set of genomes that had been originally used. 
        Then, \textit{genotype} files were produced for all the high quality GTDB representative genomes.
        Each model was then used against all the GTDB \textit{genotype} files to annotate each with the presence or the absence of the trait. 
        A list of all the phenotypic traits available for the genomes present in microbetagDB is available on \microbetag's \href{https://hariszaf.github.io/microbetag/docs/modules/phen-traits/}{documentation site}.
        The updated models are also available




    \subsection*{Pathway complementarity }
    \label{subsec:path-compl}


        To infer potential pathway complementarities we consider the modules described in KEGG MODULES database~\cite{keggmodulesdb}.
        % as functional units to highlight can provide a valid proxy.
        A KEGG module is defined as a functional unit within the KEGG framework that represents a set of enzymes and reactions involved in a specific biological process or pathway~\cite{muto2013modular}.
        Such a unit consists of several~\textit{steps}, each of which may have more than one molecular ways to occur (Figure~\ref{fig:pathcompl}).
        A module's definition is a logical expression and consists of KOs 
        % and the following symbols:
        that may be coupled with one another as:
            % a. the space, representing a connection in the pathway
            a. connected steps of the pathway
            % b. plus sign, representing a molecular complex,
            b. parts of a molecular complex, 
            % c. comma, representing alternatives and
            c. alternatives of the same step, and 
            % d. minus sign, designates an optional item in the complex.
            d. optional entities of a complex.
        Both (a) and (b) cases should be considered as the~\texttt{AND} logical operator, while (c) would be the~\texttt{OR} (Figure~\ref{fig:pathcompl}).
        Given a module's definition, we will consider as an~\textit{alternative} any subset of the KO terms mentioned in the definition, 
        that has exactly one way to perform each step, provided that all the steps of the module are covered.
        % and support the functioning of all the steps of the module.
        We define a genome as having a~\textit{complete} module, if and only if all of the KOs of at least one alternative are present on the genome.

        Within this framework,~\texttt{kofamscan}~\cite{aramaki2020kofamkoala} was used to annotate with KEGG ORTHOLOGY terms (KOs) the $16,900$ high quality GTDB representative genomes for which a~\texttt{.faa} was available~\cite{kanehisa2012kegg}.
        The KOs of each genome were then mapped to their corresponding KEGG modules; a KO may map to more than one modules ($1:n$).
        % , see Appendix~\ref{secA1}).

        \begin{figure}[h!]
            \includegraphics*[width=0.8\columnwidth]{figs/path_complem.png}
            \caption{
                Pathway complementarity approach. 
                The high quality GTDB genomes were annotated with KEGG ORTHOLOGY (KO) terms.
                The various ways of getting a KEGG module complete were enumerated and all the possible ways a donor species could "fill" a beneficiary's non-complete module were calculated.
                In this case, there are 4 unique ways for having the serine biosynthesis module complete; in all of them K00831 is required.
                However, it is missing from the beneficiary species that supports the 2 out of the 3 steps of the module's definition.
                A donor species having and potentially sharing the corresponding enzyme of K00831 may enable the beneficiary species to produce serine.
            }
            \label{fig:pathcompl}
        \end{figure}

        All module definitions were retrieved using the KEGG API and parsed to enumerate their alternatives.
        % (\href{https://github.com/hariszaf/microbetag/blob/develop/microbetagDB/mappings/kegg_mappings/parse_module_definitions.py}{\texttt{parse\_module\_definitions.py}}).
        % A dictionary was built with all the alternatives of each module. 
        %, i.e. alternative sets of KOs, for a module to be complete.
        % (\href{https://github.com/hariszaf/microbetag/blob/develop/microbetagDB/mappings/kegg_mappings/module_definition_map.json}{module\_definition\_map.json})
        Each pair of the KEGG annotated genomes was then investigated for potential pathway complementarities, 
        i.e. whether a genome lacking a number of KOs ($genome_A$) to have a complete module ($module_x$) could benefit from another's species genome(s) ($genome_B$).
        In that case, $genome_B$ does not necessarily have a complete alternative of $module_x$; as long as it has the missing KOs that $genome_A$ needs to complete an alternative of it, $genome_B$ potentially complements $genome_A$ with respect to $module_x$.
        In total, $341,568$ unique complementarities were exported.
        % (\href{https://github.com/hariszaf/microbetag/blob/develop/microbetagDB/scripts/pathway_complementarity.py}{\texttt{pathway\_complementarity.py}})

        Thanks to the graphical user interface (GUI) of the~\href{https://www.kegg.jp/kegg/docs/color_gui.html}{KEGG pathway map viewer}~\cite{kanehisa2020kegg,kanehisa2022kegg}, 
        each complementarity can be visualised as part of the closest KEGG metabolic map; 
        where the KOs contributed by the donor are shown in blue-green whereas those coming from the beneficiary genome are coloured in rose.

        \microbetag annotates the edges of a co-occurrence network by identifying pairs where both taxa map to an annotated genome present on microbetagDB.
        % Each taxon on the network has been mapped to a NCBI Taxonomy Id and their related GTDB genomes.
        Since co-occurrence networks are undirected, both nodes of a suggested association are considered as potential donors and beneficiary species. 
        When more than one GTDB representative genome map to the same NCBI Taxonomy Id all the possible genome combinations are considered.
        Finally, two edges are added in such pairs of taxa in the annotated network: 
        one considering $species_A$ as the potential beneficiary and $species_B$ as the potenial donor species, and one vice-versa. 


    \subsection*{Seed scores using genome scale metabolic reconstructions }
    \label{subsec:seeds}

        The Metabolic Complementarity Index ($MI_{Complementarity}$) measures the degree to which two microbial species can mutually assist each other by complementing each other's biosynthetic capabilities.
        As described in~\cite{phylomint_ms}, it is defined as the proportion of seed compounds of a species that can be synthesized by the metabolic network of another, but are not included in the seed set of the latter. 
        $MI_{Complementarity}$ offers an upper bound assessment of the potential for syntrophic interactions between two species.
        Further, the Metabolic Competition Index ($MI_{Competition}$) represents the similarity in two species' nutritional profiles. 
        This index establishes an upper limit on the level of competition that one species may face from another.
        Those indices have been thoroughly described and implemented in the NetCooperate~\cite{levy2015netcooperate} and NetCompt~\cite{kreimer2012netcmpt} tools correspondingly.
        We will be referring to those two indices as "seed scores".

        Recently, the~\texttt{PhyloMint}tool~\cite{phylomint_ms} was released supporting the calculation of the seed scores of GENREs in SBML format.

        In the framework of microbetag, seed scores were computed using \texttt{PhyloMint} and draft GENREs for all pairwise combinations of GTDB representative genomes that have been RAST annotated in the framework of the PATRIC database~\cite{wattam2017improvements}.
        GENREs were reconstructed using the Model SEED pipeline~\cite{henry2010high} through its Python interface \href{https://modelseedpy.readthedocs.io/en/latest/index.html}{ModelSEEDpy}.


        Moreover, the computed seed and the non-seed (i.e., set of metabolic compounds a genome can build on its own) sets of each genome were used to compute their overlap among all the pairwise combinations of those genomes.
        More specifically, the overlap of ${seed\ set}_{species_A}$ with the ${non\ seed\ set}_{species_B}$ was retrieved.
        \microbetag then annotates again the edges of the co-occurrence network where both taxa have been mapped to a at least one GTDB genome, mentioning all the KEGG maps for which there is at least one seed compound of the potentially beneficiary species 



    \subsection*{Clustering network  }
    \label{subsec:mant}

        % KF: sentence not complete
        manta is a heuristic network clustering algorithm that clusters nodes within weighted networks effectively, leveraging the presence of negative edges and discerning between weak and 
        microbetag invokes manta~\cite{rottjers2020manta} to infer clusters from the microbial network.
        A taxonomically-informed layout is 

        strong cluster assignments.
        ++ taxonomy layout 
        

        % https://ramellose.github.io/manta/demo_manta.html



    \subsection*{Groups of annotations }
    \label{groups}

        Biologically meaningful groups were built using the micrO ontology~\cite{blank2016micro}.





    \subsection*{Building the CytoscapeApp }
    \label{subsec:build-cytoapp}

        The \microbetag CytoscapeApp was build based on the~\href{https://github.com/RBVI/scNetViz}{source code} of the scVizNet~\cite{choudhary2021scnetviz}.
        Java 
        @Ermis to add 
        

        Enrichment analysis is supported. 
        Hypergeometric distribution
        FDR +++




    \subsection*{Dependencies, Web server and API }
    \label{subsec:webserver}

        The \microbetag web service is container - based and consists of three Docker~\cite{merkel2014docker} (v24.0.2) images: 
        a. the~\href{https://www.mysql.com}{MySQL} database 
        b. an nginx~\cite{nginx} web server and 
        c. the app itself. 
        The latter uses~\href{https://gunicorn.org}{Gunicorn} (20.1.0) to build an application server which communicates with the web server using the Web Server Gateway Interface (WSGI) protocol and handles incoming HTTP requests.
        % Nginx acts as a reverse proxy server that sits in front of Gunicorn and forwards client requests to Gunicorn for processing. It serves as an intermediary between clients and your application server.
        \microbetag is implemented as a~\href{https://flask.palletsprojects.com/en/3.0.x/}{Flask} application (v2.3.2); Flask is 
        a micro web framework for developing Python web applications and RESTful APIs. 
        % REST stands for Representational State Transfer 
        A thorough description of \microbetag's API is available at the \href{https://hariszaf.github.io/microbetag/docs/api/}{ReadTheDocs web site}. 
        The source code of the \microbetag web service is available on~\href{https://github.com/msysbio/microbetagApp/}{GitHub}.

        python 3.11 slim docker image 
        julia 1.7.1 for flashweave 
        mysql.connector 8.0.27 python library
        pandas 2.1.1.
        numpy 1.26.0
        multiprocessing

        text processing using awk 

        KEGG API






% ====================================================
% Results
% ====================================================
\section*{Results
    \footnote{
        Significant advance over previously published software (usually demonstrated by direct comparison with available related software)
        This should include the findings of the study including, if appropriate, results of statistical analysis which must be included either in the text or as tables and figures. 
        This section may be combined with the Discussion section for Software articles.
    }
}
\label{sec:results}



    \subsection*{\microbetag and microbetagDB}
    \label{subsec:microbetagdb}


        \begin{figure}[H]
            \label{fig:wf}
            \includegraphics[width=0.9\columnwidth]{figs/microbetag-wf.png}
            \caption{
                Diagram of the \microbetag pre - calculations and the on the fly workflow. 
                GTDB v207 representative genomes were filtered and for those of high-quality
                33 phenotypic traits were predicted using~\texttt{phenotrex}.
                To this end, models were re-trained to sync with recent version of eggNOG.
            }
        \end{figure}



        \begin{table}[ht]
            \centering
            % \caption{microbetagDB in numbers}
            \caption[Summary of Data]{Summary of Data\footnote{s involved in $>184$ million beneficiary - donor pairs' complementarities }}
            \begin{tabular}{l r}
                \toprule
                Description & Entries \\
                \midrule
                GTDB representative genomes & 34,608 \\
                Phen-model-oriented metabolic functions & 32 \\
                FAPROTAX functions & 92 \\
                Unique complement* & 341,568 \\
                GENREs leading to $\sim1$ billion competition and complementarity scores & 30,755 \\
                \bottomrule
            \end{tabular}
        \end{table}

        % \begin{table}[ht]
        %     \centering
        %     \begin{threeparttable}
        %         \caption[Summary of Data]{Summary of Data\tnote{a}}
        %         \label{tab:data_summary}
        %         \begin{tabular}{l r}
        %             \toprule
        %             Description & Value \\
        %             \midrule
        %             GTDB representative genomes & 34,608 \\
        %             Phen-model-oriented metabolic functions & 32 \\
        %             FAPROTAX functions & 92 \\
        %             Unique complements involved in $>184$ million beneficiary - donor pairs' complementarities & 341,568 \\
        %             GENREs leading to $\sim1$ billion competition and complementarity scores & 30,755 \\
        %             \bottomrule
        %         \end{tabular}
        %         \begin{tablenotes}
        %             \item[a] Note: This is a sample note for the table.
        %         \end{tablenotes}
        %     \end{threeparttable}
        % \end{table}



        annotated network returned in \texttt{.cyjs} format


        For a computationally efficient way to annotate large networks, a Docker image is provided so the user runs a taxonomy assignment using the IDTAXA algorithm~\cite{murali2018idtaxa} of the DECIPHER R package~\cite{wright2016using}.
        A co-occurrence network is also built using FlashWeave~\cite{flashweave_cite}, as \microbetag also does. 
            

        

    \subsection*{\microbetag CytoscapeApp}
    \label{subsec:cytoapp}

        Overall comment, the CytoscapeApp returns averages and s.d. for example in seed scores. 
        If you want the exact values, go through the API. 

        % -----------
        % stats table
        % -----------
        \begin{table}[ht]

            \centering

            \begin{adjustbox}{width=1.3\textwidth,center=\textwidth}

            \begin{tabular}{c|c}

                \begin{tabular}{ccc}

                    \multicolumn{3}{c}{A. GTDB-tk: 480 bins} \\
                    \toprule
                    Step &  Time(sec) & Notes \\
                    \toprule

                    Taxonomy mapping & Cell 1,2 & on the fly \\

                    Network inference & Cell 2,2 & on the fly  \\

                    microbetag annotations & Cell 3,2 & on the fly \\

                    manta clustering & Cell 4,2 & on the fly \\

                \end{tabular} &

                \begin{tabular}{ccc}

                    \multicolumn{3}{c}{B. GTDB 16S: 3000 ASVs} \\
                    \toprule
                    Step &  Time(sec) & Notes \\
                    \toprule

                    Taxonomy assignment &  & Docker image on HP\protect\footnotemark \\

                    Taxonomy mapping & Cell 1,2 & Cell 1,3 \\

                    Network inference & Cell 2,2 & Cell 2,3 \\

                    microbetag annotations & Cell 3,2 & Cell 3,3 \\

                    manta clustering & Cell 4,2 & Cell 4,3 \\

                \end{tabular} \\
                \\
                \hline
                \\
                \begin{tabular}{ccc}

                    \multicolumn{3}{c}{C. Silva: } \\
                    \toprule
                    Step &  Time(sec) & Notes \\
                    \toprule

                    Taxonomy mapping & Cell 1,2 & Cell 1,3 \\

                    Network inference & Cell 2,2 & Cell 2,3 \\

                    microbetag annotations & Cell 3,2 & Cell 3,3 \\

                    manta clustering & Cell 4,2 & Cell 4,3 \\

                \end{tabular} &

                \begin{tabular}{ccc}

                    \multicolumn{3}{c}{D. fuzzywuzzy: } \\
                    \toprule
                    Step &  Time(sec) & Notes \\
                    \toprule

                    Taxonomy mapping & Cell 1,2 & Cell 1,3 \\

                    Network inference & Cell 2,2 & Cell 2,3 \\

                    microbetag annotations & Cell 3,2 & Cell 3,3 \\

                    manta clustering & Cell 4,2 & Cell 4,3 \\

                \end{tabular} \\

            \end{tabular}

            \end{adjustbox}

            \caption{
                Computating times per step using an abundance table of 400 taxa with taxonomy: 
                A. taxonomy scheme 
                B. 
                C.
                D. 
                \protect\footnotemark[\value{footnote}] specs of the laptop used.
            }
            
            \label{tab:grid}
        \end{table}

        The app was based on the StringApp and supported by the NRNB group.


    \subsection*{Validation of \microbetag potential}
    \label{subsec:validation}


        vitamin dataset~\cite{hessler2023vitamin}

        Metagenomic or metabarcoding data are often used to predict microbial interactions in complex communities, but these predictions are rarely explored experimentally. 
        Here, we use an organism abundance correlation network to investigate factors that control community organization in mine tailings-derived laboratory microbial consortia grown under dozens of conditions.

        The network is overlaid with metagenomic information about functional capacities to generate testable hypotheses.


        Thiamine alternative pathway~\cite{llavero2022thiamine, romine2017underlying}~\todo{Study those 2 to understand our findings}



% ====================================================
% Discussion
% ====================================================
\section*{Discussion
    \footnote{
        The user interface should be described and a discussion of the intended uses of the software, and the benefits that are envisioned, should be included, together with data on how its performance and functionality compare with, and improve, on functionally similar existing software. 
        A case study of the use of the software may be presented. 
        The planned future development of new features, if any, should be mentioned.
    }
}
\label{sec:discussion}


    \subsection*{Interpreting a real-world network with \microbetag}
    \label{subsec:usecase}


        Annelies' dataset. 


    \subsection*{\microbetag as a resource }
    \label{subsec:resource}



    \subsection*{Limitations}
    \label{subsec:limitations}


        As shown in~\cite{chklovski2023checkm2} (see Figure 6b), the original version of CheckM~\cite{parks2015checkm} that is still used on GTDB returns lower completeness scores to genomes that correspond to phyla known for having shorter genomes in general, e.g. Patescibacteria representative genomes on GTDB have an average completeness \~65\%.
        \microbetag inherits this in the filtering process for getting only high quality genomes and thus, only few representatives from these taxonomic groups are present on microbetagDB.


        It is well known that higher-order interactions, i.e. interactions involving more than two species~\cite{zelezniak2015metabolic}
        Pairwise relationships do not capture more complex forms of ecological interactions, in which one species depends on (or is influenced by) multiple other species.~\cite{faust2012microbialReviewInteractions}
    




    \subsection*{Future work}
    \label{subsec:future}


        Further indices using the seed concept have been also presented such as the metabolic interaction potential ($MIP$) and the metabolic resource overlap ($MRO$).
        $MIP$ is defined as the difference between the minimal number of components required for the growth of all members in a noninteracting community and an interacting community, i.e. the maximum number of essential nutritional components that a community can provide for itself through interspecies metabolic exchanges~\cite{zelezniak2015metabolic}.
        Similarly, $MRO$ is defined as the maximum possible overlap between the minimal nutritional requirements of all member species~\cite{zelezniak2015metabolic}.
        Regression and association rule mining~\cite{assocciationRuleMining} can be applied to address this challenge.

        \begin{itemize}
            \item pathway and seed complementarities for higher-order interactions 
            \item spatial dimension 
            \item transcriptomics data integration: compare potential complementarities with what is going on 
            \item 
        \end{itemize}




% ====================================================
% Conclusions
% ====================================================
\section*{Conclusions}
\label{sec:conclusions}

    \footnote{
        This should state clearly the main conclusions and provide an explanation of the importance and relevance of the case, data, opinion, database or software reported.
    }

    Data integration 




\backmatter



\bmhead{Supplementary information}
    \label{supplementary-files}
    \footnote{
        If your article has accompanying supplementary file(s) please state so here.
        E.g. supplementary figures and tables captions. 
    }



% ====================================================
% Declarations
% ====================================================
\section*{Declarations}


    \begin{itemize}

        \item \textbf{Availability of data and materials}

            \begin{itemize}
                \item Raw sequences for the use case: 
                \item Raw data for the validations case:
            \end{itemize}

        \item \textbf{Funding}
        
            This work was initiated thanks to an EMBO Scientific Exchange Grant to HZ. 
            It was then supported by the 3D’omics Horizon project (101000309). 
            We would also like to thank the National Resource for Network Biology (NRNB) and the Google Summer of Code 2023 for the support of E.I.M.D.

        \item \textbf{Conflict of interest/Competing interests} 

            The authors declare that they have no other competing interests.

        \item \textbf{Authors' contributions}
            \footnote{Based on the \href{https://www-elsevier-com.kuleuven.e-bronnen.be/researcher/author/policies-and-guidelines/credit-author-statement}{CRediT system}. Current list is indicative.}

            Conceptualization: K.F.
            Methodology: K.F. and H.Z.
            Software: H.Z., E.I.M.D. and J.M
            Validation: H.Z. and K.F.
            Formal analysis: H.Z. and K.F.
            Investigation: H.Z.
            Resources: K.F., A.E. and A.G.
            Data Curation: H.Z.
            Writing - Original Draft: H.Z. and K.F. 
            Writing - Review \& Editing: all
            Visualization: H.Z.
            Supervision: K.F., H.Z. and S.M.
            Project administration: K.F.
            Funding acquisition: K.F., H.Z.


        \item \textbf{Acknowledgements}

            We would like to thank Dr Christina Pavloudi and ++ for the insight on how to organise the trait groups.


        \item \textbf{Ethics approval}

            Not applicable

        \item \textbf{Consent to participate}
        
            Not applicable.

        \item \textbf{Code availability: }

            \begin{itemize}
                \item microbetagDB related scripts: \href{https://github.com/hariszaf/microbetag}{https://github.com/hariszaf/microbetag}
                \item microbetagApp and webserver: \href{https://github.com/msysbio/microbetagApp}{https://github.com/msysbio/microbetagApp}.
                \item CytoscapeApp: \href{https://github.com/ermismd/MGG/}{https://github.com/ermismd/MGG/}
                \item Validation and use case: <think of having that under the 3D'omics organization> 
                \item Documentation web-site: \href{https://hariszaf.github.io/microbetag/}{https://hariszaf.github.io/microbetag/}
            \end{itemize}

    \end{itemize}


% ====================================================
% Appendices
% ====================================================
\begin{appendices}

    \section{Mappings}
    \label{secA1}
        $n:1$
        $n:n$ etc

    \section{Background on seed scores and complementarities}
    \label{secA2}

        \subsection{Background on seed scores}

            In that case, once a seed is assured, it activates all the rest of that group.
            Therefore, a confidence level ($C$) ranging from 0 to 1, has been previously described to quantify the relevance of each seed:

            \begin{equation}
                \label{eq:confidence}
                C_i = 1 / seed's\ group\ with \ i \ size
            \end{equation}

            $C=0$ corresponds to a non-seed node, while $C=1$ represents an independent node. 



            \begin{equation}
                \label{eq:mi_complementarity}
                MI_{Complementarity} = \frac{ | SeedSet_A \cap \neg SeedSet_B | }{ | SeedSet_A \cap (SeedSet_B \cup \neg SeedSet_B) | }
            \end{equation}


            As also described in~\cite{phylomint_ms}, it is calculated as the proportion of compounds in a species' seed set that coincide with those in an other's, while also factoring in the confidence scores associated with seed compounds.

            \begin{equation}
                \label{eq:mi_competition}
                MI_{Competition} = \frac{ \sum C (SeedSet_A \cap SeedSet_B) }{ \sum{ C (SeedSet_A )} }
            \end{equation}


        \subsection{Background on pathway complementarity}

            For example, the definition of the D-Galacturonate degradation in Bacteria (\href{https://www.genome.jp/dbget-bin/www_bget?M00631}{M00631}) is: \vspace{0.35cm}


            \makebox[1in][l]{K01812 K00041 (K01685,K16849+K16850) K00874 (K01625,K17463)}
            
            % \vspace{0.25cm}
            \begin{flushleft}
                that once breaking down, it leads to 4 alternative sets of KOs (pathways):\vspace{0.25cm}
            \end{flushleft} 

            K01812 K00041 K01685 K00874 K01625\vspace{0.25cm}

            K01812 K00041 K16849+K16850 K00874 K01625\vspace{0.25cm}

            K01812 K00041 K01685 K00874 K17463\vspace{0.25cm}

            K01812 K00041 K16849+K16850 K00874 K17463\vspace{0.5cm}


        \subsection{Complementarities}
            
            KEGG compound 
            ModelSEED compounds
            ModelSEED compounds mapped to KEGG compounds and kept only those related to KEGG modules. 


\end{appendices}


% ======================
% Bibliography
% ======================
\bibliography{sn-bibliography}


\end{document}








% ----------------------- TABLES ----------------------
% \section{Tables}\label{sec5}

% Tables can be inserted via the normal table and tabular environment. To put
% footnotes inside tables you should use \verb+\footnotetext[]{...}+ tag.
% The footnote appears just below the table itself (refer Tables~\ref{tab1} and \ref{tab2}). 
% For the corresponding footnotemark use \verb+\footnotemark[...]+

% \begin{table}[h]
% \caption{Caption text}\label{tab1}%
% \begin{tabular}{@{}llll@{}}
% \toprule
% Column 1 & Column 2  & Column 3 & Column 4\\
% \midrule
% row 1    & data 1   & data 2  & data 3  \\
% row 2    & data 4   & data 5\footnotemark[1]  & data 6  \\
% row 3    & data 7   & data 8  & data 9\footnotemark[2]  \\
% \botrule
% \end{tabular}
% \footnotetext{Source: This is an example of table footnote. This is an example of table footnote.}
% \footnotetext[1]{Example for a first table footnote. This is an example of table footnote.}
% \footnotetext[2]{Example for a second table footnote. This is an example of table footnote.}
% \end{table}

% \noindent
% The input format for the above table is as follows:

% %%=============================================%%
% %% For presentation purpose, we have included  %%
% %% \bigskip command. please ignore this.       %%
% %%=============================================%%
% \bigskip
% \begin{verbatim}
% \begin{table}[<placement-specifier>]
% \caption{<table-caption>}\label{<table-label>}%
% \begin{tabular}{@{}llll@{}}
% \toprule
% Column 1 & Column 2 & Column 3 & Column 4\\
% \midrule
% row 1 & data 1 & data 2	 & data 3 \\
% row 2 & data 4 & data 5\footnotemark[1] & data 6 \\
% row 3 & data 7 & data 8	 & data 9\footnotemark[2]\\
% \botrule
% \end{tabular}
% \footnotetext{Source: This is an example of table footnote. 
% This is an example of table footnote.}
% \footnotetext[1]{Example for a first table footnote.
% This is an example of table footnote.}
% \footnotetext[2]{Example for a second table footnote. 
% This is an example of table footnote.}
% \end{table}
% \end{verbatim}
% \bigskip
% %%=============================================%%
% %% For presentation purpose, we have included  %%
% %% \bigskip command. please ignore this.       %%
% %%=============================================%%

% \begin{table}[h]
% \caption{Example of a lengthy table which is set to full textwidth}\label{tab2}
% \begin{tabular*}{\textwidth}{@{\extracolsep\fill}lcccccc}
% \toprule%
% & \multicolumn{3}{@{}c@{}}{Element 1\footnotemark[1]} & \multicolumn{3}{@{}c@{}}{Element 2\footnotemark[2]} \\\cmidrule{2-4}\cmidrule{5-7}%
% Project & Energy & $\sigma_{calc}$ & $\sigma_{expt}$ & Energy & $\sigma_{calc}$ & $\sigma_{expt}$ \\
% \midrule
% Element 3  & 990 A & 1168 & $1547\pm12$ & 780 A & 1166 & $1239\pm100$\\
% Element 4  & 500 A & 961  & $922\pm10$  & 900 A & 1268 & $1092\pm40$\\
% \botrule
% \end{tabular*}
% \footnotetext{Note: This is an example of table footnote. This is an example of table footnote this is an example of table footnote this is an example of~table footnote this is an example of table footnote.}
% \footnotetext[1]{Example for a first table footnote.}
% \footnotetext[2]{Example for a second table footnote.}
% \end{table}

% \vfill\eject

% In case of double column layout, tables which do not fit in single column width should be set to full text width. For this, you need to use \verb+\begin{table*}+ \verb+...+ \verb+\end{table*}+ instead of \verb+\begin{table}+ \verb+...+ \verb+\end{table}+ environment. Lengthy tables which do not fit in textwidth should be set as rotated table. For this, you need to use \verb+\begin{sidewaystable}+ \verb+...+ \verb+\end{sidewaystable}+ instead of \verb+\begin{table*}+ \verb+...+ \verb+\end{table*}+ environment. This environment puts tables rotated to single column width. For tables rotated to double column width, use \verb+\begin{sidewaystable*}+ \verb+...+ \verb+\end{sidewaystable*}+.

% \begin{sidewaystable}
% \caption{Tables which are too long to fit, should be written using the ``sidewaystable'' environment as shown here}\label{tab3}
% \begin{tabular*}{\textheight}{@{\extracolsep\fill}lcccccc}
% \toprule%
% & \multicolumn{3}{@{}c@{}}{Element 1\footnotemark[1]}& \multicolumn{3}{@{}c@{}}{Element\footnotemark[2]} \\\cmidrule{2-4}\cmidrule{5-7}%
% Projectile & Energy	& $\sigma_{calc}$ & $\sigma_{expt}$ & Energy & $\sigma_{calc}$ & $\sigma_{expt}$ \\
% \midrule
% Element 3 & 990 A & 1168 & $1547\pm12$ & 780 A & 1166 & $1239\pm100$ \\
% Element 4 & 500 A & 961  & $922\pm10$  & 900 A & 1268 & $1092\pm40$ \\
% Element 5 & 990 A & 1168 & $1547\pm12$ & 780 A & 1166 & $1239\pm100$ \\
% Element 6 & 500 A & 961  & $922\pm10$  & 900 A & 1268 & $1092\pm40$ \\
% \botrule
% \end{tabular*}
% \footnotetext{Note: This is an example of table footnote this is an example of table footnote this is an example of table footnote this is an example of~table footnote this is an example of table footnote.}
% \footnotetext[1]{This is an example of table footnote.}
% \end{sidewaystable}




% ------------------   FIGURES ---------------------

% \section{Figures}\label{sec6}

% As per the \LaTeX\ standards you need to use eps images for \LaTeX\ compilation and \verb+pdf/jpg/png+ images for \verb+PDFLaTeX+ compilation. This is one of the major difference between \LaTeX\ and \verb+PDFLaTeX+. Each image should be from a single input .eps/vector image file. Avoid using subfigures. The command for inserting images for \LaTeX\ and \verb+PDFLaTeX+ can be generalized. The package used to insert images in \verb+LaTeX/PDFLaTeX+ is the graphicx package. Figures can be inserted via the normal figure environment as shown in the below example:

% %%=============================================%%
% %% For presentation purpose, we have included  %%
% %% \bigskip command. please ignore this.       %%
% %%=============================================%%
% \bigskip
% \begin{verbatim}
% \begin{figure}[<placement-specifier>]
% \centering
% \includegraphics{<eps-file>}
% \caption{<figure-caption>}\label{<figure-label>}
% \end{figure}
% \end{verbatim}
% \bigskip
% %%=============================================%%
% %% For presentation purpose, we have included  %%
% %% \bigskip command. please ignore this.       %%
% %%=============================================%%

% \begin{figure}[h]%
% \centering
% \includegraphics[width=0.9\textwidth]{fig.eps}
% \caption{This is a widefig. This is an example of long caption this is an example of long caption  this is an example of long caption this is an example of long caption}\label{fig1}
% \end{figure}

% In case of double column layout, the above format puts figure captions/images to single column width. To get spanned images, we need to provide \verb+\begin{figure*}+ \verb+...+ \verb+\end{figure*}+.

% For sample purpose, we have included the width of images in the optional argument of \verb+\includegraphics+ tag. Please ignore this. 




% ----------------  ALGORITHSM ------------------------

% Packages \verb+algorithm+, \verb+algorithmicx+ and \verb+algpseudocode+ are used for setting algorithms in \LaTeX\ using the format:

% %%=============================================%%
% %% For presentation purpose, we have included  %%
% %% \bigskip command. please ignore this.       %%
% %%=============================================%%
% \bigskip
% \begin{verbatim}
% \begin{algorithm}
% \caption{<alg-caption>}\label{<alg-label>}
% \begin{algorithmic}[1]
% . . .
% \end{algorithmic}
% \end{algorithm}
% \end{verbatim}
% \bigskip
% %%=============================================%%
% %% For presentation purpose, we have included  %%
% %% \bigskip command. please ignore this.       %%
% %%=============================================%%

% You may refer above listed package documentations for more details before setting \verb+algorithm+ environment. For program codes, the ``verbatim'' package is required and the command to be used is \verb+\begin{verbatim}+ \verb+...+ \verb+\end{verbatim}+. 

% Similarly, for \verb+listings+, use the \verb+listings+ package. \verb+\begin{lstlisting}+ \verb+...+ \verb+\end{lstlisting}+ is used to set environments similar to \verb+verbatim+ environment. Refer to the \verb+lstlisting+ package documentation for more details.

% A fast exponentiation procedure:

% \lstset{texcl=true,basicstyle=\small\sf,commentstyle=\small\rm,mathescape=true,escapeinside={(*}{*)}}
% \begin{lstlisting}
% begin
%   for $i:=1$ to $10$ step $1$ do
%       expt($2,i$);  
%       newline() od                (*\textrm{Comments will be set flush to the right margin}*)
% where
% proc expt($x,n$) $\equiv$
%   $z:=1$;
%   do if $n=0$ then exit fi;
%      do if odd($n$) then exit fi;                 
%         comment: (*\textrm{This is a comment statement;}*)
%         $n:=n/2$; $x:=x*x$ od;
%      { $n>0$ };
%      $n:=n-1$; $z:=z*x$ od;
%   print($z$). 
% end
% \end{lstlisting}

% \begin{algorithm}
% \caption{Calculate $y = x^n$}\label{algo1}
% \begin{algorithmic}[1]
% \Require $n \geq 0 \vee x \neq 0$
% \Ensure $y = x^n$ 
% \State $y \Leftarrow 1$
% \If{$n < 0$}\label{algln2}
%         \State $X \Leftarrow 1 / x$
%         \State $N \Leftarrow -n$
% \Else
%         \State $X \Leftarrow x$
%         \State $N \Leftarrow n$
% \EndIf
% \While{$N \neq 0$}
%         \If{$N$ is even}
%             \State $X \Leftarrow X \times X$
%             \State $N \Leftarrow N / 2$
%         \Else[$N$ is odd]
%             \State $y \Leftarrow y \times X$
%             \State $N \Leftarrow N - 1$
%         \EndIf
% \EndWhile
% \end{algorithmic}
% \end{algorithm}



% ----------------- EQUATIONS --------------------

% \section{Equations}\label{sec4}

% Equations in \LaTeX\ can either be inline or on-a-line by itself (``display equations''). For
% inline equations use the \verb+$...$+ commands. E.g.: The equation
% $H\psi = E \psi$ is written via the command \verb+$H \psi = E \psi$+.

% For display equations (with auto generated equation numbers)
% one can use the equation or align environments:
% \begin{equation}
% \|\tilde{X}(k)\|^2 \leq\frac{\sum\limits_{i=1}^{p}\left\|\tilde{Y}_i(k)\right\|^2+\sum\limits_{j=1}^{q}\left\|\tilde{Z}_j(k)\right\|^2 }{p+q}.\label{eq1}
% \end{equation}
% where,
% \begin{align}
% D_\mu &=  \partial_\mu - ig \frac{\lambda^a}{2} A^a_\mu \nonumber \\
% F^a_{\mu\nu} &= \partial_\mu A^a_\nu - \partial_\nu A^a_\mu + g f^{abc} A^b_\mu A^a_\nu \label{eq2}
% \end{align}
% Notice the use of \verb+\nonumber+ in the align environment at the end
% of each line, except the last, so as not to produce equation numbers on
% lines where no equation numbers are required. The \verb+\label{}+ command
% should only be used at the last line of an align environment where
% \verb+\nonumber+ is not used.
% \begin{equation}
% Y_\infty = \left( \frac{m}{\textrm{GeV}} \right)^{-3}
%     \left[ 1 + \frac{3 \ln(m/\textrm{GeV})}{15}
%     + \frac{\ln(c_2/5)}{15} \right]
% \end{equation}
% The class file also supports the use of \verb+\mathbb{}+, \verb+\mathscr{}+ and
% \verb+\mathcal{}+ commands. As such \verb+\mathbb{R}+, \verb+\mathscr{R}+
% and \verb+\mathcal{R}+ produces $\mathbb{R}$, $\mathscr{R}$ and $\mathcal{R}$
% respectively (refer Subsubsection~\ref{subsubsec2}).
