<!DOCTYPE html><html lang="en-US"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=Edge"><link rel="stylesheet" href="/assets/css/just-the-docs-default.css"><style type="text/css"> .site-nav > .nav-list:nth-child(1):not(.nav-category-list) > .nav-list-item:not(.external):nth-child(2) > .nav-list-link { display: block; font-weight: 600; text-decoration: none; background-image: linear-gradient(-90deg, #ebedf5 0%, rgba(235, 237, 245, 0.8) 80%, rgba(235, 237, 245, 0) 100%); } .site-nav > .nav-list:nth-child(1):not(.nav-category-list) > .nav-list-item:nth-child(2) > .nav-list-expander svg { transform: rotate(-90deg); } .site-nav > .nav-list:nth-child(1):not(.nav-category-list) > .nav-list-item:nth-child(2) > .nav-list { display: block; } .site-nav > .nav-category-list > .nav-list-item > .nav-list-expander svg { transform: rotate(-90deg); } .site-nav > .nav-category-list > .nav-list-item > .nav-list { display: block; }</style><script src="/assets/js/vendor/lunr.min.js"></script> <script src="/assets/js/just-the-docs.js"></script><meta name="viewport" content="width=device-width, initial-scale=1"><link rel="icon" href="/favicon.ico" type="image/x-icon"><title>Modules | microbetag</title><meta name="generator" content="Jekyll v4.3.2" /><meta property="og:title" content="Modules" /><meta property="og:locale" content="en_US" /><meta name="description" content="A place with the background and the how to of the microbetag tool - currently in production" /><meta property="og:description" content="A place with the background and the how to of the microbetag tool - currently in production" /><link rel="canonical" href="http://localhost:4000/docs/modules" /><meta property="og:url" content="http://localhost:4000/docs/modules" /><meta property="og:site_name" content="microbetag" /><meta property="og:type" content="website" /><meta name="twitter:card" content="summary" /><meta property="twitter:title" content="Modules" /> <script type="application/ld+json"> {"@context":"https://schema.org","@type":"WebPage","description":"A place with the background and the how to of the microbetag tool - currently in production","headline":"Modules","url":"http://localhost:4000/docs/modules"}</script> <script src="https://polyfill.io/v3/polyfill.js?features=es6"> </script> <script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script><body> <a class="skip-to-main" href="#main-content">Skip to main content</a> <svg xmlns="http://www.w3.org/2000/svg" class="d-none"> <symbol id="svg-link" viewBox="0 0 24 24"><title>Link</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-link"><path d="M10 13a5 5 0 0 0 7.54.54l3-3a5 5 0 0 0-7.07-7.07l-1.72 1.71"></path><path d="M14 11a5 5 0 0 0-7.54-.54l-3 3a5 5 0 0 0 7.07 7.07l1.71-1.71"></path> </svg> </symbol> <symbol id="svg-menu" viewBox="0 0 24 24"><title>Menu</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-menu"><line x1="3" y1="12" x2="21" y2="12"></line><line x1="3" y1="6" x2="21" y2="6"></line><line x1="3" y1="18" x2="21" y2="18"></line> </svg> </symbol> <symbol id="svg-arrow-right" viewBox="0 0 24 24"><title>Expand</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-chevron-right"><polyline points="9 18 15 12 9 6"></polyline> </svg> </symbol> <symbol id="svg-external-link" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-external-link"><title id="svg-external-link-title">(external link)</title><path d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"></path><polyline points="15 3 21 3 21 9"></polyline><line x1="10" y1="14" x2="21" y2="3"></line> </symbol> <symbol id="svg-doc" viewBox="0 0 24 24"><title>Document</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-file"><path d="M13 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V9z"></path><polyline points="13 2 13 9 20 9"></polyline> </svg> </symbol> <symbol id="svg-search" viewBox="0 0 24 24"><title>Search</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-search"> <circle cx="11" cy="11" r="8"></circle><line x1="21" y1="21" x2="16.65" y2="16.65"></line> </svg> </symbol> <symbol id="svg-copy" viewBox="0 0 16 16"><title>Copy</title><svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="bi bi-clipboard" viewBox="0 0 16 16"><path d="M4 1.5H3a2 2 0 0 0-2 2V14a2 2 0 0 0 2 2h10a2 2 0 0 0 2-2V3.5a2 2 0 0 0-2-2h-1v1h1a1 1 0 0 1 1 1V14a1 1 0 0 1-1 1H3a1 1 0 0 1-1-1V3.5a1 1 0 0 1 1-1h1v-1z"/><path d="M9.5 1a.5.5 0 0 1 .5.5v1a.5.5 0 0 1-.5.5h-3a.5.5 0 0 1-.5-.5v-1a.5.5 0 0 1 .5-.5h3zm-3-1A1.5 1.5 0 0 0 5 1.5v1A1.5 1.5 0 0 0 6.5 4h3A1.5 1.5 0 0 0 11 2.5v-1A1.5 1.5 0 0 0 9.5 0h-3z"/> </svg> </symbol> <symbol id="svg-copied" viewBox="0 0 16 16"><title>Copied</title><svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="bi bi-clipboard-check-fill" viewBox="0 0 16 16"><path d="M6.5 0A1.5 1.5 0 0 0 5 1.5v1A1.5 1.5 0 0 0 6.5 4h3A1.5 1.5 0 0 0 11 2.5v-1A1.5 1.5 0 0 0 9.5 0h-3Zm3 1a.5.5 0 0 1 .5.5v1a.5.5 0 0 1-.5.5h-3a.5.5 0 0 1-.5-.5v-1a.5.5 0 0 1 .5-.5h3Z"/><path d="M4 1.5H3a2 2 0 0 0-2 2V14a2 2 0 0 0 2 2h10a2 2 0 0 0 2-2V3.5a2 2 0 0 0-2-2h-1v1A2.5 2.5 0 0 1 9.5 5h-3A2.5 2.5 0 0 1 4 2.5v-1Zm6.854 7.354-3 3a.5.5 0 0 1-.708 0l-1.5-1.5a.5.5 0 0 1 .708-.708L7.5 10.793l2.646-2.647a.5.5 0 0 1 .708.708Z"/> </svg> </symbol> </svg><div class="side-bar"><div class="site-header" role="banner"> <a href="/" class="site-title lh-tight"> microbetag </a> <button id="menu-button" class="site-button btn-reset" aria-label="Toggle menu" aria-pressed="false"> <svg viewBox="0 0 24 24" class="icon" aria-hidden="true"><use xlink:href="#svg-menu"></use></svg> </button></div><nav aria-label="Main" id="site-nav" class="site-nav"><ul class="nav-list"><li class="nav-list-item"><a href="/" class="nav-list-link">Home</a><li class="nav-list-item"><button class="nav-list-expander btn-reset" aria-label="toggle items in Modules category" aria-pressed="false"> <svg viewBox="0 0 24 24" aria-hidden="true"><use xlink:href="#svg-arrow-right"></use></svg> </button><a href="/docs/modules" class="nav-list-link">Modules</a><ul class="nav-list"><li class="nav-list-item"><a href="/docs/modules/faprotax-functions/" class="nav-list-link">FAPROTAX functions</a><li class="nav-list-item"><a href="/docs/modules/phen-traits/" class="nav-list-link">Traits predicted based on phenDB models</a></ul><li class="nav-list-item"><a href="/docs/cytoApp/" class="nav-list-link">Cytoscape App</a><li class="nav-list-item"><a href="/docs/api/" class="nav-list-link">API documentation</a><li class="nav-list-item"><a href="/docs/use_case/" class="nav-list-link">Use case</a></ul><ul class="nav-list"><li class="nav-list-item external"> <a href="https://github.com/hariszaf/microbetag/tree/gh-pages" class="nav-list-link external"> microbetag docs on GitHub <svg viewBox="0 0 24 24" aria-labelledby="svg-external-link-title"><use xlink:href="#svg-external-link"></use></svg> </a></ul></nav><footer class="site-footer"> This site uses <a href="https://github.com/just-the-docs/just-the-docs">Just the Docs</a>, a documentation theme for Jekyll.</footer></div><div class="main" id="top"><div id="main-header" class="main-header"><div class="search" role="search"><div class="search-input-wrap"> <input type="text" id="search-input" class="search-input" tabindex="0" placeholder="Search microbetag" aria-label="Search microbetag" autocomplete="off"> <label for="search-input" class="search-label"><svg viewBox="0 0 24 24" class="search-icon"><use xlink:href="#svg-search"></use></svg></label></div><div id="search-results" class="search-results"></div></div><nav aria-label="Auxiliary" class="aux-nav"><ul class="aux-nav-list"><li class="aux-nav-list-item"> <a href="https://github.com/hariszaf/microbetag" class="site-button" > microbetag on GitHub </a></ul></nav></div><div class="main-content-wrap"><div id="main-content" class="main-content"><main><h1 class="no_toc" id="modules"> <a href="#modules" class="anchor-heading" aria-labelledby="modules"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Modules</h1><h2 class="no_toc text-delta" id="table-of-contents"> <a href="#table-of-contents" class="anchor-heading" aria-labelledby="table-of-contents"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Table of contents</h2><ol id="markdown-toc"><li><a href="#functional-annotations" id="markdown-toc-functional-annotations">Functional annotations</a><ol><li><a href="#based-on-faprotax" id="markdown-toc-based-on-faprotax">Based on FAPROTAX</a><li><a href="#based-on-phendb" id="markdown-toc-based-on-phendb">Based on phenDB</a></ol><li><a href="#pathway-complementarity" id="markdown-toc-pathway-complementarity">Pathway complementarity</a><li><a href="#seed-scores-based-on-genome-scale-draft-reconstructions-gems" id="markdown-toc-seed-scores-based-on-genome-scale-draft-reconstructions-gems">Seed scores based on genome-scale draft reconstructions (GEMs)</a><li><a href="#references" id="markdown-toc-references">References</a></ol><hr /><p>microbetag gets as input either a co-occurrence network or an abundance table where either <a href="https://www.arb-silva.de">Silva</a> or <a href="https://gtdb.ecogenomic.org">GTDB</a> taxonomies have been used. When an abundance table is provided, microbetag firsts builds a co-occurrence network using <a href="https://github.com/meringlab/FlashWeave.jl">FlashWeave</a> [1].</p><p>Once a network is availalbe, microbetag identifies the taxonomic level that has been assigned to each entry, for example <code class="language-plaintext highlighter-rouge">D_0__Bacteria; D_1__Firmicutes; D_2__Clostridia; D_3__Clostridiales; D_4__Ruminococcaceae; D_5__uncultured; D_6__uncultured rumen bacterium</code> has reached the family level, while <code class="language-plaintext highlighter-rouge">D_0__Bacteria; D_1__Actinobacteria; D_2__Coriobacteriia; D_3__Coriobacteriales; D_4__Coriobacteriaceae; D_5__Collinsella; D_6__uncultured bacterium</code> is at the genus level and proceeds with the network annotation.</p><p>The network annotation consists of 4 major modules:</p><ul><li><strong>literature oriented</strong> taxa functional annotation using <a href="https://pages.uoregon.edu/slouca/LoucaLab/archive/FAPROTAX/lib/php/index.php"><strong>FAPROTAX</strong></a> [2]<li><strong>genomic oriented</strong> taxa functional annotation using an updated, local instance of <a href="https://phendb.org"><strong>phenDB</strong></a> using all representative genomes of GTDB and <a href="https://phenotrex.readthedocs.io/en/latest/usage.html"><code class="language-plaintext highlighter-rouge">phenotrex</code></a><li><strong>pathway complementarity</strong> annotations between taxa that have been found co-correlated in the produced (or user provided) network; both taxa were considered as potential donor and beneficiary (see <a href="#pathway-complementarity">Pathway complementarity</a> for more)<li><strong>complementarity</strong> [3] and <strong>competition</strong> [4] <strong>seed scores</strong> between draft metabolic reconstructions of GTDB representative genomes, mapped to the input taxa using <a href="https://github.com/mgtools/PhyloMint"><strong>PhyloMInt</strong></a> (see <a href="">Seed-based complementarity and competition scores</a> for more)</ul><p><strong>Nodes</strong> that have species or strain taxonomic annotation are mapped to their closest representative GTDB genomes and based on those, they get phenDB-like and FAPROTAX functional annotations. Taxa (nodes) that have been taxonomically annotated at the family or order level are annotated using FAPROTAX.</p><p><strong>Edges</strong> linking nodes that have been taxonomically assigned at the species or strain level, i.e. both nodes of the association have a species/strain taxonomic annotation, are annotated using the pathway complementarity and the seed scores approaches.</p><p>Below, you will find further background and examples of each annotation type.</p><h2 id="functional-annotations"> <a href="#functional-annotations" class="anchor-heading" aria-labelledby="functional-annotations"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Functional annotations</h2><h3 id="based-on-faprotax"> <a href="#based-on-faprotax" class="anchor-heading" aria-labelledby="based-on-faprotax"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Based on FAPROTAX</h3><p><a href="https://pages.uoregon.edu/slouca/LoucaLab/archive/FAPROTAX/lib/php/index.php"><strong>FAPROTAX</strong></a> [2] maps taxa (e.g. genera or species) to metabolic or other ecologically relevant functions based on the literature for cultured representatives. It currently comprises more than 7600 annotation rules, covering ~4700 prokaryotic clades. Each annotation rule comes with literature citations and can, thus, be independently verified. Similar 16S rRNA oriented approaches (e.g., PICRUSt, Tax4Fun etc.) estimate community gene content based on available sequenced genomes. On the contrary, FAPROTAX estimates metabolic phenotypes based on experimental evidence.</p><p>The taxonomy assigned to each OTU/ASV (amplicon data) or bin (shotgun data) on the abundance table provided by the user, is mapped to a list of functions one can check <a href="faprotax-functions.md">here</a>.</p><p>As an example, here is how the FAPROTAX output looks like for the <strong>denitrification</strong> function for three samples (columns):</p><p><img src="../../assets/images/faprotax_denitrification.png" alt="faprotax example denitrif" /></p><p>FAPROTAX returns only the ASVs present in the (user provided) abundance table that are related to the <strong>denetrification</strong> function. Numbers represent the ASV abundance in each sample.</p><p>microbetag runs FAPROTAX agains the abundance table and parses the subtables (<a href="https://github.com/msysbio/microbetagApp/blob/main/services/web/microbetag/scripts/utils.py#L181"><code class="language-plaintext highlighter-rouge">seqId_faprotax_functions_assignment</code></a>) to annotate each node with the corresponding function.</p><p>In case the user provides as input a co-occurrence network, microbetag runs FAPROTAX against the nodes.</p><h3 id="based-on-phendb"> <a href="#based-on-phendb" class="anchor-heading" aria-labelledby="based-on-phendb"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Based on phenDB</h3><p><a href="https://phenotrex.readthedocs.io/en/latest/usage.html"><code class="language-plaintext highlighter-rouge">phenotrex</code></a> enables phenotypic trait prediction on user’s metagenomic genomes/bins.</p><p>Phenotrex classifiers were re-trained using the genomes provided by phenDB for each model. For example, for the acetic acid production case, the <a href="https://phendb.org/reports/modeldetails?model_id=16">corresponding webpage of phenDB</a> pointed to the set of genomes that had been originally used. These genomes were recovered and the classifiers were re-trained to sync with the latest version of <a href="http://eggnog5.embl.de/#/app/home">eggNOG</a>.</p><p>Under the <a href="phen-traits.md">Traits predicted based on phenDB models</a> tab, we provide a description of each feature abbreviation returned from microbetag based on phen feautures, based on those from the <a href="https://phendb.org/reports/modeloverview">phenDB group</a>.</p><p>The annotation is referring to the species under study. Each trait gets a “Yes” or “No” decision along with an accurracy score. For example <code class="language-plaintext highlighter-rouge">NOB</code> : <em>species under study is part of the clade of NOB</em>.</p><p>Here is an example of how two GTDB genomes look like:</p><p><img src="../../assets/images/phen_traits_fmt.png" alt="phen traits example" /></p><p>microbetag annotates all network nodes (corresponding to OTUs/ASVs/bins that have been identified to species/strain level) mapped to a representative GTDB genome with these functional traits and scores.</p><h2 id="pathway-complementarity"> <a href="#pathway-complementarity" class="anchor-heading" aria-labelledby="pathway-complementarity"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Pathway complementarity</h2><p>As defined by the <a href="https://www.genome.jp/kegg/module.html">KEGG resource</a>, <em>“the KEGG MODULE database is a manually curated collection of modular functional units, categorized into pathway modules, signature modules and reaction modules”</em>.</p><p>All the GTDB representative genomes were KEGG annottated. Considering all pair-wised combinatons of those genomes, microbetag checks whether the KEGG Orthology (KO) terms of a genome (donor) could complete a KEGG module of another (beneficary), if shared.</p><p>Here is an example where Acidiferrobacter sp. SPIII3 (GCA_003184265.1) potentially shares K01626 to complete the Shikimate pathway (M00022) of Prochlorococcus marinus AS9601 (GCA_000015645.1).</p><p><img src="../../assets/images/kegg_example.png" alt="complementarity in kegg example" /></p><p>As several genomes can be mapped to the same NCBI Taxonomy id, microbetag returns all possible complementarities between all the donor’s and the beneficary’s genomes.</p><p>microbetag annotates all <strong>edges</strong> where both nodes represent species/strain level taxonomies with such complementarities.</p><h2 id="seed-scores-based-on-genome-scale-draft-reconstructions-gems"> <a href="#seed-scores-based-on-genome-scale-draft-reconstructions-gems" class="anchor-heading" aria-labelledby="seed-scores-based-on-genome-scale-draft-reconstructions-gems"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Seed scores based on genome-scale draft reconstructions (GEMs)</h2><p>Based on Borenstein <em>et al.</em> (2008) [5] a metabolic network’s “seed set” is the set of compounds that, based on the network topology, are exogenously acquired”. Here is an example (based on the <a href="http://borensteinlab.com/software_netseed_examples.html">Borensteil lab webpage</a>):</p><p><img src="../../assets/images/seed_concept_example.png" alt="seed concept" /></p><p>Node A is a seed, as it cannot be activated by any other node in the network. Nodes F, G, and H are also seeds but they are interdependent, i.e. activating one of these nodes would activate the rest, but at least one must be active to activate the rest. These nodes form a “seed group”. To quantify the relevance of each identified seed, we assign each seed a <strong>confidence level (C)</strong>, ranging from 0 to 1. A confidence level of 0 would correspond to a non-seed node, while a 1 would correspond to a seed that cannot be activated by another node. Seeds which belong to a seed group with more than 1 seed are given a fractional confidence level, the inverse of the number of seeds in the group.</p><p>Based on the seed concept, several scores between metabolic models of pair of species have been described. As described in the PhyloMInt paper: the <strong>Metabolic Complementarity Index</strong> (\(MI_{Complementarity}\)) is calculated as the <em>“fraction of the seed set of the genome-scale reconstruction of species A, that is found within B’s metabolic network but not part of B’s seed set, normalized by the number of A’s seed set in B’s entire metabolic network”</em>. This complementarity score represents the <strong>potential for A’s to utilize the potential metabolic output of B</strong>.</p>\[MI_{Complementarity} = \frac {\lvert SeedSetA \cap \neg SeedSetB \rvert} { SeedSetA \cap (SeedSetB \cup \neg SeeedSetB)}\]<p>Similarly, as described in the PhyloMInt paper, the <strong>Metabolic Competition Index</strong> (\(MI_{Competition}\)) <em>“is calculated as the fraction of A’s seed set that is also in B’s seed set, normalized by the weighted sum of the confidence score”</em>. MIC estimates the baseline <strong>metabolic overlap</strong> between two given metabolic networks.</p>\[MI_{Competition} = \frac {\sum C( SeedSetA \cap SeedSetB )} {\sum C(SeedSetA)}\]<p>Here is a toy example to calculate the two indices as shown in the PhyloMInt paper:</p><p><img src="../../assets/images/seed-scores-examples.png" alt="seed scores example" /> In a comparison between metabolic network A versus metabolic network B, metabolic network A shares only one seed metabolite with metabolic network B (metabolite F) which lies in the seed group in metabolic network A. Thus, the \(MI_{Competition}\) between metabolic network A and B is \((1/3) / 2 = 1/6\).</p><p class="highlight">The \(1/3\) term represents the confidence level of the seed group node.</p><p>Among SeedSetA, metabolites A and F are found within the metabolic network B but only metabolite A is within non-SeedSetB, thus the MIComplementarity index between metabolic network A and metabolic network B is 0.5.</p><p>These indexes can be used in various types of metabolic networks. In the framework of microbetag, all GTDB representative genomes were used to come up with draft genome-scale reconstructions using <a href="https://github.com/ModelSEED/ModelSEEDpy"><code class="language-plaintext highlighter-rouge">modelseedpy</code></a> with its default gapfilling algorithm and a complete medium. Then, all GEMs pair-wised combinations were considered and using <a href="https://github.com/mgtools/PhyloMint"><code class="language-plaintext highlighter-rouge">PhyloMInt</code></a> their \(MI_{Complementarity}\) and \(MI_{Competition}\) scores were calculated. microbetag annotates all <strong>edges</strong> between species/strain level taxonomically assigned nodes with such scores, considering all the representative GTDB genomes mapping to the corresponding NCBI Taxonomy ids of the nodes.</p><h2 id="references"> <a href="#references" class="anchor-heading" aria-labelledby="references"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> References</h2><p>[1] Tackmann, J., Rodrigues, J.F.M. and von Mering, C., 2019. Rapid inference of direct interactions in large-scale ecological networks from heterogeneous microbial sequencing data. Cell systems, 9(3), pp.286-296, DOI: <a href="https://doi.org/10.1016/j.cels.2019.08.002">10.1016/j.cels.2019.08.002</a>.</p><p>[2] Louca, S., Parfrey, L.W., Doebeli, M. (2016) - Decoupling function and taxonomy in the global ocean microbiome. Science 353:1272-1277, DOI: <a href="https://doi.org/10.1126/science.aaf4507">10.1126/science.aaf4507</a>.</p><p>[3] Levy, R., Carr, R., Kreimer, A., Freilich, S., Borenstein, E. “NetCooperate: a network-based tool for inferring host-microbe and microbe-microbe cooperation.” BMC Bioinformatics, 2015.</p><p>[4] Kreimer, A., Doron-Faigenboim, A., Borenstein, E., Freilich, S. “NetCmpt: a network-based tool for calculating the metabolic competition between bacterial species.” Bioinformatics, 2012.</p><p>[5] Borenstein, E., Kupiec, M., Feldman, M.W. and Ruppin, E., 2008. Large-scale reconstruction and phylogenetic analysis of metabolic environments. Proceedings of the National Academy of Sciences, 105(38), pp.14482-14487.</p><hr><h2 class="text-delta">Table of contents</h2><ul><li> <a href="/docs/modules/faprotax-functions/">FAPROTAX functions</a><li> <a href="/docs/modules/phen-traits/">Traits predicted based on phenDB models</a></ul></main><hr><footer><p><a href="#top" id="back-to-top">Back to top</a></p><p class="text-small text-grey-dk-100 mb-0">Copyright &copy; 2017-2020 Patrick Marsceill. Distributed by an <a href="https://github.com/just-the-docs/just-the-docs/tree/main/LICENSE.txt">MIT license.</a> <a href="https://www.netlify.com/">This site is powered by Netlify.</a></p><div class="d-flex mt-2"><p class="text-small text-grey-dk-000 mb-0"> <a href="https://github.com/just-the-docs/just-the-docs/tree/main/docs/modules/modules.md" id="edit-this-page">Edit this page on GitHub</a></p></div></footer></div></div><div class="search-overlay"></div></div><script src="https://cdn.jsdelivr.net/npm/mermaid@9.1.6/dist/mermaid.min.js"></script> <script> var config = {} ; mermaid.initialize(config); window.mermaid.init(undefined, document.querySelectorAll('.language-mermaid')); </script>
